lib/IDETool/CompilerInvocation.cpp

```diff
@@ -83,18 +84,16 @@ static FrontendInputsAndOutputs resolveSymbolicLinksInInputs(
     std::string &Error) {
   assert(FileSystem);
 
-  llvm::SmallString<128> PrimaryFile;
-  if (auto err = FileSystem->getRealPath(UnresolvedPrimaryFile, PrimaryFile))
-    PrimaryFile = UnresolvedPrimaryFile;
+  std::string PrimaryFile = resolveSymbolicLinks(
+      UnresolvedPrimaryFile, FileSystem.get());
 
   unsigned primaryCount = 0;
   // FIXME: The frontend should be dealing with symlinks, maybe similar to
   // clang's FileManager ?
   FrontendInputsAndOutputs replacementInputsAndOutputs;
   for (const InputFile &input : inputsAndOutputs.getAllInputs()) {
-    llvm::SmallString<128> newFilename;
-    if (auto err = FileSystem->getRealPath(input.getFileName(), newFilename))
-      newFilename = input.getFileName();
+    llvm::SmallString<128> newFilename = StringRef(resolveSymbolicLinks(
+        input.getFileName(), FileSystem.get()));
```

Comment:
Hmmm this part is unfortunate, it's the only caller that needs a SmallString and all just for the `native` call :(

Reply:
I could overload the function to have one overload that outputs to a `SmallVectorBase&` and one that returns an `std::string`? I'm open to suggestions.

Reply:
Alternatively, we could add in a new optional parameter to `resolveSymbolicLinks`:

```
resolveSymbolicLinks(StringRef, llvm::vfs::FileSystem *, std::optional<llvm::sys::path::Style> = std::nullopt);
```

If the style is requested, we convert the path to the particular style before returning.

Reply:
That works for me too. I'm happy with either suggestion here.

Reply:
I implemented both, in that I implemented @compnerd 's solution and it lead me to privately use a static overload taking a `SmallVectorImpl`. See latest revision.